macros.c3

```diff
@@ -0,0 +1,72 @@
+module test;
+import std::io;
+import std::math;
+import std::collections::list;
+
+struct Point2
+{
+	float x, y;
+}
+
+// Short function syntax
+fn float Point2.at(Point2 this, int i) @operator([]) => i ? this.x : this.y;
+fn int Point2.len(Point2 this) => 2;
+
+macro bool is_float_vec($Type)
+{
+	var $temp = $Type{};
+	$if $checks($temp[0]):
+		$switch
+			$case (@typekind($temp[0]) != FLOAT):
+				return false;
+			$case $checks($temp.len()):
+				return values::@is_int($temp.len());
+			$case $checks($temp.len):
+				return values::@is_int($temp.len);
+			$default:
+				return false;
+		$endswitch
+	$else
+		return false;
+	$endif
+}
+
+// Use contracts rather than asserts:
+
+/**
+ * @require is_float_vec($typeof(vec))
+ **/
+macro norm(vec)
+{
+	var $Type = $typeof(vec[0]);
+	// We have some types with `vec.len` and some with `vec.len()`
+	$if $checks(vec.len()):
+		var size = vec.len();
+	$else
+		var size = vec.len;
+	$endif
+	if (!size) return 0;
+	$Type result = vec[0] * vec[0];
+	for ($typeof(size) i = 1; i < size; i++)
+	{
+		result += vec[i] * vec[i];
+	}
+	return math::sqrt(result);
+}
+
+fn float norm2(float x, float y) => norm(Point2{x, y});
```

Comment:
Does `norm2` compile down behind the scenes to the hardcoded solution?

Reply:
Yes. So with -O1 this is the LLVM IR:
```llvm
; Function Attrs: nofree nosync nounwind memory(none)
define float @test.norm2(float %0, float %1) local_unnamed_addr #1 {
entry:
  %fmul = fmul float %1, %1
  %fmul1 = fmul float %0, %0
  %fadd = fadd float %fmul, %fmul1
  %2 = tail call float @llvm.sqrt.f32(float %fadd)
  ret float %2
}
```
Creating a hand coded function:
```c
fn float norm2a(float x, float y)
{
  Point2 a = {x, y};
  int size = a.len();
  if (!size) return 0;
  float result = a[0] * a[0];
  for (int i = 1; i < size; i++)
  {
    result += a[i] * a[i];
  }
  return math::sqrt(result);
}
```
Yields this at -O1 (I only use -O1 because the -O0 is so much longer)

```llvm
; Function Attrs: nofree nosync nounwind memory(none)
define float @test.norm2a(float %0, float %1) local_unnamed_addr #1 {
entry:
  %fmul = fmul float %1, %1
  %fmul1 = fmul float %0, %0
  %fadd = fadd float %fmul, %fmul1
  %2 = tail call float @llvm.sqrt.f32(float %fadd)
  ret float %2
}
```

Reply:
It's possible to special case the non-dynamic length, but I didn't do this in this example. Should I?

Reply:
I just meant like in my commented forms for norm2. I don't have active code for the simpler form in any of the languages. I already merged this pr, but if you want to make a new one with a commented hardcoded 2d form, that's fine:

https://github.com/contextfreecode/abstractions/blob/8090c29641994e09616fc78725c4a14bd3373268/comptime.zig#L49-L52

Reply:
Neither Rust nor Zig is doing that on zero optimization. So maybe it's better to leave the version as it is since any trivial optimization will fix it anyway.